File: src/scanner.ts
import fs from "node:fs/promises";
import path from "node:path";
import { spawnSync } from "node:child_process";

export type PackageManager = "yarn" | "pnpm" | "npm" | "bun" | "unknown";

export type RepoScan = {
  repoPath: string;
  nodeModulesPath: string;
  manager: PackageManager;
  hasLockfile: boolean;
  inactiveDays: number | null;
  bytes: number;
  reinstallCommand: string;
  git: {
    branch: string | null;
    dirty: boolean;
    ahead: number;
    behind: number;
  };
};

export type ScanProgress = {
  directoriesScanned: number;
};

type ScanOptions = {
  onProgress?: (progress: ScanProgress) => void;
  onRepoFound?: (row: RepoScan) => void;
  onRepoUpdated?: (row: RepoScan) => void;
  progressInterval?: number;
  sizeConcurrency?: number;
  ignoreDirs?: string[];
};

const SKIP_DIRS = new Set([
  "node_modules",
  ".git",
  ".next",
  ".nuxt",
  ".turbo",
  "dist",
  "build",
  "coverage",
  "out",
  "target"
]);

const getGitMeta = (repoPath: string): RepoScan["git"] => {
  const branchResult = spawnSync("git", ["-C", repoPath, "branch", "--show-current"], { encoding: "utf8" });
  const branch = branchResult.status === 0 ? branchResult.stdout.trim() || null : null;

  const dirtyResult = spawnSync("git", ["-C", repoPath, "status", "--porcelain"], { encoding: "utf8" });
  const dirty = dirtyResult.status === 0 ? dirtyResult.stdout.trim().length > 0 : false;

  const aheadBehindResult = spawnSync(
    "git",
    ["-C", repoPath, "rev-list", "--left-right", "--count", "@{upstream}...HEAD"],
    { encoding: "utf8" }
  );
  if (aheadBehindResult.status !== 0) {
    return { branch, dirty, ahead: 0, behind: 0 };
  }
  const [behindRaw, aheadRaw] = aheadBehindResult.stdout.trim().split(/\s+/);
  const ahead = Number(aheadRaw ?? 0);
  const behind = Number(behindRaw ?? 0);
  return {
    branch,
    dirty,
    ahead: Number.isFinite(ahead) ? ahead : 0,
    behind: Number.isFinite(behind) ? behind : 0
  };
};

const detectManager = async (repoPath: string): Promise<{ manager: PackageManager; hasLockfile: boolean }> => {
  const locks: Array<{ file: string; manager: PackageManager }> = [
    { file: "yarn.lock", manager: "yarn" },
    { file: "pnpm-lock.yaml", manager: "pnpm" },
    { file: "package-lock.json", manager: "npm" },
    { file: "bun.lockb", manager: "bun" },
    { file: "bun.lock", manager: "bun" }
  ];

  for (const lock of locks) {
    try {
      await fs.access(path.join(repoPath, lock.file));
      return { manager: lock.manager, hasLockfile: true };
    } catch {
      // ignore
    }
  }

  return { manager: "unknown", hasLockfile: false };
};

const getReinstallCommand = (manager: PackageManager): string => {
  switch (manager) {
    case "yarn":
      return "yarn install --immutable";
    case "pnpm":
      return "pnpm install --frozen-lockfile";
    case "npm":
      return "npm ci";
    case "bun":
      return "bun install --frozen-lockfile";
    default:
      return "install command unknown";
  }
};

const isGitRepository = (repoPath: string): boolean => {
  const result = spawnSync("git", ["-C", repoPath, "rev-parse", "--is-inside-work-tree"], { encoding: "utf8" });
  return result.status === 0 && result.stdout.trim() === "true";
};

const getInactiveDays = (repoPath: string): number | null => {
  const result = spawnSync("git", ["-C", repoPath, "log", "-1", "--format=%ct"], { encoding: "utf8" });
  if (result.status !== 0) return null;
  const ts = Number(result.stdout.trim());
  if (!Number.isFinite(ts) || ts <= 0) return null;
  const days = (Date.now() / 1000 - ts) / 86400;
  return Math.max(0, Math.floor(days));
};

const getDirectorySize = async (dir: string): Promise<number> => {
  const du = spawnSync("du", ["-sk", dir], { encoding: "utf8" });
  if (du.status === 0) {
    const kb = Number(du.stdout.trim().split(/\s+/)[0]);
    if (Number.isFinite(kb) && kb >= 0) {
      return kb * 1024;
    }
  }

  let total = 0;
  const stack = [dir];
  while (stack.length) {
    const current = stack.pop()!;
    let entries;
    try {
      entries = await fs.readdir(current, { withFileTypes: true });
    } catch {
      continue;
    }
    for (const entry of entries) {
      const full = path.join(current, entry.name);
      if (entry.isDirectory()) {
        stack.push(full);
      } else if (entry.isFile()) {
        try {
          const stat = await fs.stat(full);
          total += stat.size;
        } catch {
          // ignore
        }
      }
    }
  }
  return total;
};

const createLimiter = (concurrency: number) => {
  let active = 0;
  const queue: Array<() => void> = [];

  const next = () => {
    active -= 1;
    const run = queue.shift();
    if (run) run();
  };

  return async <T>(task: () => Promise<T>): Promise<T> => {
    if (active >= concurrency) {
      await new Promise<void>((resolve) => queue.push(resolve));
    }
    active += 1;
    try {
      return await task();
    } finally {
      next();
    }
  };
};

export const findReposWithNodeModules = async (root: string, options: ScanOptions = {}): Promise<RepoScan[]> => {
  const scans: RepoScan[] = [];
  const queue = [root];
  let queueIndex = 0;
  let directoriesScanned = 0;
  const progressInterval = Math.max(1, options.progressInterval ?? 100);
  let lastReported = 0;
  const runSizeTask = createLimiter(Math.max(1, options.sizeConcurrency ?? 6));
  const sizeTasks: Array<Promise<void>> = [];
  const skipDirs = new Set([...SKIP_DIRS, ...(options.ignoreDirs ?? [])]);

  while (queueIndex < queue.length) {
    const current = queue[queueIndex]!;
    queueIndex += 1;
    directoriesScanned += 1;
    if (directoriesScanned === 1 || directoriesScanned - lastReported >= progressInterval) {
      lastReported = directoriesScanned;
      options.onProgress?.({ directoriesScanned });
    }
    let entries;
    try {
      entries = await fs.readdir(current, { withFileTypes: true });
    } catch {
      continue;
    }

    const names = new Set(entries.map((entry) => entry.name));
    const hasGitMarker = names.has(".git");
    const hasPackageJson = names.has("package.json");
    const hasNodeModules = names.has("node_modules");

    if (hasPackageJson && hasNodeModules && (hasGitMarker || isGitRepository(current))) {
      const nodeModulesPath = path.join(current, "node_modules");
      const { manager, hasLockfile } = await detectManager(current);
      const row: RepoScan = {
        repoPath: current,
        nodeModulesPath,
        manager,
        hasLockfile,
        inactiveDays: getInactiveDays(current),
        bytes: 0,
        reinstallCommand: getReinstallCommand(manager),
        git: getGitMeta(current)
      };
      scans.push(row);
      options.onRepoFound?.(row);

      sizeTasks.push(
        runSizeTask(async () => {
          row.bytes = await getDirectorySize(nodeModulesPath);
          options.onRepoUpdated?.(row);
        })
      );
      continue;
    }

    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      if (entry.name.startsWith(".")) continue;
      if (skipDirs.has(entry.name)) continue;
      queue.push(path.join(current, entry.name));
    }
  }

  if (directoriesScanned !== lastReported) {
    options.onProgress?.({ directoriesScanned });
  }
  await Promise.all(sizeTasks);

  return scans;
};

File: internal/scanner/scanner.go
package scanner

import (
	"fmt"
	"io/fs"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"

	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/plumbing"
)

type Manager string

const (
	ManagerYarn    Manager = "yarn"
	ManagerPnpm    Manager = "pnpm"
	ManagerNpm     Manager = "npm"
	ManagerBun     Manager = "bun"
	ManagerUnknown Manager = "unknown"
)

type GitContext struct {
	Branch string `json:"branch"`
	Dirty  bool   `json:"dirty"`
	Ahead  int    `json:"ahead"`
	Behind int    `json:"behind"`
}

type Candidate struct {
	RepoPath         string     `json:"repoPath"`
	NodeModulesPath  string     `json:"nodeModulesPath"`
	Manager          Manager    `json:"manager"`
	HasLockfile      bool       `json:"hasLockfile"`
	InactiveDays     *int       `json:"inactiveDays"`
	Bytes            int64      `json:"bytes"`
	ReinstallCommand string     `json:"reinstallCommand"`
	Git              GitContext `json:"git"`
}

type Progress struct {
	DirectoriesScanned int
}

type Options struct {
	OnProgress     func(Progress)
	OnFound        func(Candidate)
	OnSized        func(Candidate)
	ProgressEvery  int
	SizeConcurrent int
	IgnoreDirs     []string
}

var defaultSkip = map[string]struct{}{
	"node_modules": {}, ".git": {}, ".next": {}, ".nuxt": {},
	".turbo": {}, "dist": {}, "build": {}, "coverage": {},
	"out": {}, "target": {},
}

func Scan(root string, opts Options) ([]Candidate, error) {
	if opts.ProgressEvery <= 0 {
		opts.ProgressEvery = 100
	}
	if opts.SizeConcurrent <= 0 {
		opts.SizeConcurrent = 6
	}

	skip := make(map[string]struct{}, len(defaultSkip)+len(opts.IgnoreDirs))
	for k := range defaultSkip {
		skip[k] = struct{}{}
	}
	for _, d := range opts.IgnoreDirs {
		skip[d] = struct{}{}
	}

	var (
		mu       sync.Mutex
		out      []Candidate
		queue    = []string{filepath.Clean(root)}
		qIdx     int
		dirCount int
		lastRep  int
		lim      = make(chan struct{}, opts.SizeConcurrent)
		wg       sync.WaitGroup
	)

	reportProgress := func() {
		if dirCount == 1 || dirCount-lastRep >= opts.ProgressEvery {
			lastRep = dirCount
			if opts.OnProgress != nil {
				opts.OnProgress(Progress{DirectoriesScanned: dirCount})
			}
		}
	}

	for qIdx < len(queue) {
		current := queue[qIdx]
		qIdx++
		dirCount++
		reportProgress()

		entries, err := os.ReadDir(current)
		if err != nil {
			continue
		}

		var hasGit, hasPkg, hasNM bool
		for _, e := range entries {
			switch e.Name() {
			case ".git":
				if e.IsDir() {
					hasGit = true
				}
			case "package.json":
				if !e.IsDir() {
					hasPkg = true
				}
			case "node_modules":
				if e.IsDir() {
					hasNM = true
				}
			}
		}

		if hasPkg && hasNM && (hasGit || isGitWorkTree(current)) {
			row := buildCandidate(current)
			mu.Lock()
			out = append(out, row)
			mu.Unlock()
			if opts.OnFound != nil {
				opts.OnFound(row)
			}

			wg.Add(1)
			lim <- struct{}{}
			go func(r Candidate, idx int) {
				defer wg.Done()
				defer func() { <-lim }()
				size, _ := dirSize(r.NodeModulesPath)
				mu.Lock()
				out[idx].Bytes = size
				updated := out[idx]
				mu.Unlock()
				if opts.OnSized != nil {
					opts.OnSized(updated)
				}
			}(row, len(out)-1)
			continue
		}

		for _, e := range entries {
			if !e.IsDir() || strings.HasPrefix(e.Name(), ".") {
				continue
			}
			if _, skipIt := skip[e.Name()]; skipIt {
				continue
			}
			queue = append(queue, filepath.Join(current, e.Name()))
		}
	}

	if dirCount != lastRep && opts.OnProgress != nil {
		opts.OnProgress(Progress{DirectoriesScanned: dirCount})
	}
	wg.Wait()
	return out, nil
}

func buildCandidate(repoPath string) Candidate {
	nm := filepath.Join(repoPath, "node_modules")
	mgr, hasLock := detectManager(repoPath)
	return Candidate{
		RepoPath:         repoPath,
		NodeModulesPath:  nm,
		Manager:          mgr,
		HasLockfile:      hasLock,
		InactiveDays:     inactiveDays(repoPath),
		ReinstallCommand: reinstallCommand(mgr),
		Git:              gitContext(repoPath),
	}
}

func isGitWorkTree(dir string) bool {
	repo, err := git.PlainOpen(dir)
	if err != nil {
		return false
	}
	_, err = repo.Head()
	return err == nil
}

func detectManager(repoPath string) (Manager, bool) {
	locks := []struct {
		file string
		mgr  Manager
	}{
		{"yarn.lock", ManagerYarn},
		{"pnpm-lock.yaml", ManagerPnpm},
		{"package-lock.json", ManagerNpm},
		{"bun.lockb", ManagerBun},
		{"bun.lock", ManagerBun},
	}
	for _, l := range locks {
		if fileExists(filepath.Join(repoPath, l.file)) {
			return l.mgr, true
		}
	}
	return ManagerUnknown, false
}

func reinstallCommand(m Manager) string {
	switch m {
	case ManagerYarn:
		return "yarn install --immutable"
	case ManagerPnpm:
		return "pnpm install --frozen-lockfile"
	case ManagerNpm:
		return "npm ci"
	case ManagerBun:
		return "bun install --frozen-lockfile"
	default:
		return "install command unknown"
	}
}

func gitContext(repoPath string) GitContext {
	repo, err := git.PlainOpen(repoPath)
	if err != nil {
		return GitContext{}
	}
	head, err := repo.Head()
	if err != nil {
		return GitContext{}
	}
	branch := head.Name().Short()
	if branch == "" {
		branch = head.Name().String()
	}

	wt, err := repo.Worktree()
	dirty := false
	if err == nil {
		st, err := wt.Status()
		dirty = err == nil && !st.IsClean()
	}

	ahead, behind := aheadBehind(repo, head)
	return GitContext{Branch: branch, Dirty: dirty, Ahead: ahead, Behind: behind}
}

func aheadBehind(repo *git.Repository, head *plumbing.Reference) (ahead, behind int) {
	name := head.Name().Short()
	if name == "" {
		return 0, 0
	}
	branch, err := repo.Branch(name)
	if err != nil || branch == nil || branch.Merge == "" {
		return 0, 0
	}
	mergeRef := branch.Merge.Short()
	if mergeRef == "" {
		mergeRef = branch.Merge.String()
	}
	remoteRef := plumbing.NewRemoteReferenceName(branch.Remote, mergeRef)
	upstream, err := repo.Reference(remoteRef, true)
	if err != nil {
		return 0, 0
	}
	ahead, _ = countReachableUntil(repo, head.Hash(), upstream.Hash())
	behind, _ = countReachableUntil(repo, upstream.Hash(), head.Hash())
	return ahead, behind
}

func countReachableUntil(repo *git.Repository, from, until plumbing.Hash) (int, error) {
	if from == until {
		return 0, nil
	}
	iter, err := repo.Log(&git.LogOptions{From: from})
	if err != nil {
		return 0, err
	}
	defer iter.Close()
	count := 0
	for {
		c, err := iter.Next()
		if err != nil {
			break
		}
		if c.Hash == until {
			return count, nil
		}
		count++
	}
	return count, nil
}

func inactiveDays(repoPath string) *int {
	repo, err := git.PlainOpen(repoPath)
	if err != nil {
		return nil
	}
	head, err := repo.Head()
	if err != nil {
		return nil
	}
	commit, err := repo.CommitObject(head.Hash())
	if err != nil {
		return nil
	}
	when := commit.Committer.When
	if when.IsZero() {
		when = commit.Author.When
	}
	days := int(time.Since(when).Hours() / 24)
	if days < 0 {
		days = 0
	}
	return &days
}

func dirSize(dir string) (int64, error) {
	if runtime.GOOS == "darwin" || runtime.GOOS == "linux" {
		out, err := exec.Command("du", "-sk", dir).Output()
		if err == nil {
			fields := strings.Fields(string(out))
			if len(fields) > 0 {
				var kb int64
				if _, err := fmt.Sscanf(fields[0], "%d", &kb); err == nil && kb >= 0 {
					return kb * 1024, nil
				}
			}
		}
	}

	var total int64
	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			return nil
		}
		info, err := d.Info()
		if err != nil {
			return nil
		}
		total += info.Size()
		return nil
	})
	return total, err
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
